### PR TITLE
Auto-migrate Okta users to new groups on startup + new update/delete-user self-restrictions + remove ability to update user email

### PR DESCRIPTION
**Pull request by jeremyzitomer-usds**
## Related Issue or Background Info

- Fulfills the backfill step of https://app.zenhub.com/workspaces/simplereport-5ff8be8854ae8b000e378210/issues/cdcgov/prime-simplereport/700
- Makes it such that a user cannot update their own role/privileges or delete themselves
- Please ignore the Sonar failure, this (obviously) can't be tested anywhere but in the LiveOktaRepository

### CODE REVIEW

**Comment by jeremyzitomer-usds** [comment]
accidentally merged partial changes into this branch from a downstream branch regarding self-restrictions on udpate/delete roles. 

however, these changes as-is create a problem where updateUser() now can't be called to update a user's own name/contact information, which is resolved in a downstream branch in changes that haven't yet been merged to this branch. 

for that reason, i'm going to merge the facility-access-control changes into this branch when they are ready, and add them to this PR.

_however, this PR as-is currently isolates how the Okta group migration process will happen, so it may be valuable to look at now, before the new changes for proper access control are added._

**Comment by jeremyzitomer-usds** [comment]
if you saw a comment from me on this PR about a downstream mix-up, ignore it :) PR is ready to review as-is and is self-contained.

(again, please ignore Sonar failure.)

**Comment by jeremyzitomer-usds** [comment]
if you saw a recent comment from me on this PR about merging from a downstream branch, ignore it. all issues resolved, but now, as the title of this PR suggests, this PR has not just Okta migration, but also a small change to access control restrictions when trying to update/delete one's own user.

this is ready for review. ignore Sonar failure as discussed above.

**Comment by jeremyzitomer-usds** [comment]
In a fun turn of events, and after merging up the base branch and seeing a suspicious failing `updateUser_self_success()` test, I have now realized that updating one's own email invalidates how all of their demo and live authentication works, since the IdentitySupplier results get mangled. 

and with that, much to the joy of @benwarfield-usds i'm sure, i have removed the capability to update a user's email. there's no FE concerns afaik @nickscialli-usds since the FE hasn't wired in the feature to update a user's name/contact info yet.

File: backend/src/main/java/gov/cdc/usds/simplereport/config/AuthorizationConfiguration.java [hunk 1]
```diff
@@ -30,13 +30,22 @@
...
           + ".userHasPermission("
           + "T(gov.cdc.usds.simplereport.config.authorization.UserPermission).";
+
+  private static final String SPEL_IS_SITE_ADMIN =
+      "@" + AUTHORIZER_BEAN + ".userHasSiteAdminRole()";
+
+  private static final String SPEL_IS_NOT_SELF = "@" + AUTHORIZER_BEAN + ".userIsNotSelf(#userId)";
+
+  private static final String SPEL_IS_NOT_IN_SAME_ORG =
+      "@" + AUTHORIZER_BEAN + ".userIsInSameOrg(#userId)";
```
Comment (jeskew-gov) [inline backend/src/main/java/gov/cdc/usds/simplereport/config/AuthorizationConfiguration.java:40]: Shouldn't this be named `SPEL_IS_IN_SAME_ORG`? This tripped me up on [line 67](https://github.com/CDCgov/prime-simplereport/pull/1037/files#diff-b5255ef1f759cdfb6352cd1e0a6ca5d1df0e8230844a97464a997e4aa09d5fb8R67)

File: backend/src/main/java/gov/cdc/usds/simplereport/config/AuthorizationConfiguration.java [hunk 2]
```diff
@@ -47,19 +56,42 @@
...
+   * Require the current user to to be one of the administrative users ("superusers") or have the
+   * {@link UserPermission#MANAGE_USERS} permission for the organization containing user with UUID
+   * {@code userId}; and require the user to not be the user they are operating on.
+   *
+   * <p>NOTE: any method with this annotation must have a parameter {@code userId}.
+   */
+  @Retention(RUNTIME)
+  @Target(METHOD)
+  @PreAuthorize(
+      "( "
```
Comment (jeskew-gov) [inline backend/src/main/java/gov/cdc/usds/simplereport/config/AuthorizationConfiguration.java:81]: I'm usually not a fan of expression languages due to how tough they make composing complex expressions from simple ones, and this one is pretty hard to follow.

I don't think this is actionable in this PR, but maybe Spring has a way to make this type of expression composition more readable? I think a good first step would be replacing the `SPEL_HAS_PERMISSION` static string with a method that takes a UserPermission and returns a complete clause, but that may not be compatible with Java annotation processing.

File: backend/src/main/java/gov/cdc/usds/simplereport/db/model/ApiUser.java [hunk 3]
```diff
@@ -16,7 +16,7 @@
 @DynamicUpdate
 public class ApiUser extends EternalSystemManagedEntity {
 
-  @Column(nullable = false, updatable = true, unique = true)
+  @Column(nullable = false, updatable = false, unique = true)
   @NaturalId(mutable = true)
```
Comment (jeskew-gov) [inline backend/src/main/java/gov/cdc/usds/simplereport/db/model/ApiUser.java:20]: Should `mutable` still be true here?

File: backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java [hunk 4]
```diff
@@ -294,21 +293,111 @@ public void setUserIsActive(String username, Boolean active) {
...
+
+        if (migration && migrationDestRoles.contains(role)) {
+          migrationDestGroupIds.add(g.getId());
+          migrationDestGroupNames.add(roleGroupName);
+        }
+      } catch (ResourceException e) {
+        if (migration) {
+          migrationSourceGroupNames.add(roleGroupName);
+        } else {
+          // ignore attempts to create groups that already exist if this is in migration mode
```
Comment (jeskew-gov) [inline backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java:332]: Should this comment be moved to the `if` clause? I'm guessing that the call to `migrationSourceGroupNames.add(roleGroupName);` is what constitutes ignoring the exception.

File: backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java [hunk 5]
```diff
@@ -294,21 +293,111 @@ public void setUserIsActive(String username, Boolean active) {
...
+
+        if (migration) {
+          LOG.info("Created Okta group={} via migration", facilityGroupName);
+        }
+
+      } catch (ResourceException e) {
+        if (migration) {
+          migrationSourceGroupNames.add(facilityGroupName);
+        } else {
+          // ignore attempts to create groups that already exist if this is in migration mode
```
Comment (jeskew-gov) [inline backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java:358]: ditto

File: backend/src/main/java/gov/cdc/usds/simplereport/config/AuthorizationConfiguration.java (hunk 2 above)
Comment (jeremyzitomer-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/config/AuthorizationConfiguration.java:81]: yeah it's not compatible because the expression has to be a constant :\

File: backend/src/main/java/gov/cdc/usds/simplereport/config/AuthorizationConfiguration.java (hunk 1 above)
Comment (jeremyzitomer-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/config/AuthorizationConfiguration.java:40]: oopsie

File: backend/src/main/java/gov/cdc/usds/simplereport/config/AuthorizationConfiguration.java (hunk 2 above)
Comment (jeskew-gov) [reply backend/src/main/java/gov/cdc/usds/simplereport/config/AuthorizationConfiguration.java:81]: Well, I guess we're stuck with it then. Maybe in a future PR these complex predicates could be converted to single method calls on the authorizer bean? I'm not sure if there's anything beyond that we could do.

File: backend/src/test/java/gov/cdc/usds/simplereport/api/ApiUserManagementTest.java [hunk 6]
```diff
@@ -827,6 +862,17 @@ void setUserIsDeleted_orgUser_failure() {
     runQuery("set-user-is-deleted", deleteVariables, ACCESS_ERROR);
   }
 
+  @Test
+  void setUserIsDeleted_self_failure() {
```
Comment (benwarfield-usds) [inline backend/src/test/java/gov/cdc/usds/simplereport/api/ApiUserManagementTest.java:866]: 👏

**Review by benwarfield-usds — COMMENTED** [review]
Haven't gotten to the meaty part but wanted to comment on the SpEL convo quickly.

File: backend/src/main/java/gov/cdc/usds/simplereport/config/AuthorizationConfiguration.java (hunk 2 above)
Comment (benwarfield-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/config/AuthorizationConfiguration.java:81]: Agreed that moving the complex predicates into explicit code of our own is probably the medium or long-term solution.

File: backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java [hunk 7]
```diff
@@ -294,21 +293,111 @@ public void setUserIsActive(String username, Boolean active) {
...
+      try {
+        Group g =
+            GroupBuilder.instance()
+                .setName(facilityGroupName)
+                .setDescription(facilityGroupDescription)
+                .buildAndCreate(_client);
+        _app.createApplicationGroupAssignment(g.getId());
+
+        if (migration) {
+          LOG.info("Created Okta group={} via migration", facilityGroupName);
```
Comment (benwarfield-usds) [inline backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java:351]: Wouldn't we want to log the facility group creation anyway?

File: backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java [hunk 8]
```diff
@@ -294,21 +293,111 @@ public void setUserIsActive(String username, Boolean active) {
...
+
+        LOG.info(
+            "Migrated users from Okta source_groups={} to destination_groups={}",
+            migrationSourceGroupNames,
+            migrationDestGroupNames);
+      } catch (ResourceException e) {
+        LOG.error(
+            "Error migrating users from Okta source_groups={} to destination_groups={}: "
+                + "check if conflicting group_rule={} already exists. "
+                + "Did you delete groups without deleting their relevant group rules?",
```
Comment (benwarfield-usds) [inline backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java:387]: Should we schedule a clean-up to axe those rules at some point?

File: backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/OktaRepository.java [hunk 9]
```diff
@@ -32,6 +32,9 @@
 
   public Map<String, OrganizationRoleClaims> getAllUsersForOrganization(Organization org);
 
+  public void createOrganization(
```
Comment (benwarfield-usds) [inline backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/OktaRepository.java:35]: Here or in the LiveOktaRepository implementation of this method would seem like it would have been a good place to explain what happens when `migration` is `true`. The code isn't unreasonably difficult to read (subject to the constraints of the API) but it's nice to be able to cross-reference against a block comment when trying to figure out if it does what it's supposed to.

**Review by benwarfield-usds — COMMENTED** [review]
I'm concerned that we don't have a rule in here for `if user has ENTRY_ONLY then remove them from USER`. We could do that manually, but otherwise we're presumably promoting all ENTRY_ONLY users, unless I missed something.

File: backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java [hunk 10]
```diff
@@ -294,21 +293,111 @@ public void setUserIsActive(String username, Boolean active) {
...
   }
 
-  public void createOrganization(Organization org) {
+  public void createOrganization(
+      Organization org, Collection<Facility> facilities, boolean migration) {
     String name = org.getOrganizationName();
     String externalId = org.getExternalId();
+
+    // After the first migration, the migration code below will amount to a no-op,
+    // until we change the next line to new OrganizationRole's
```
[the remark this answers is not in the document]
Comment (jeremyzitomer-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java:302]: going to decline unless we can come up with something that doesn't conflict with a separate class name

File: backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java [hunk 11]
```diff
@@ -294,21 +293,111 @@ public void setUserIsActive(String username, Boolean active) {
...
+
+        if (migration) {
+          LOG.info("Created Okta group={} via migration", roleGroupName);
+        }
+
+        if (migration && migrationDestRoles.contains(role)) {
+          migrationDestGroupIds.add(g.getId());
+          migrationDestGroupNames.add(roleGroupName);
+        }
+      } catch (ResourceException e) {
```
[the remark this answers is not in the document]
Comment (jeremyzitomer-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java:328]: not without doing an exact string match on the error message, which i would very much like to not do :-\

[the remark this answers is not in the document]
Comment (jeremyzitomer-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java:328]: but if we have an invalid API token, e.g., the constructor of this class will throw an exception anyway

File: backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (benwarfield-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java:302]: Solid point.

File: backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (benwarfield-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java:328]: Yeah, that specific error case is not one that worries me too much, but errors that are actually unexpected are also swept under this rug. Ugh.

File: backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java (hunk 8 above)
Comment (jeremyzitomer-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java:387]: here's the amazing wonderful kicker -- if you delete them programmatically, it undoes all the user migration that those rules performed. if you delete them manually on the Okta UI instead, the migrated users stay where they are. so i'm planning to delete them manually following each stage of the rollout. (if you're not crazed about this idea, i'm going to shrug and point you to Okta.)

**Comment by jeremyzitomer-usds** [comment]
Okta gives us an easy way to _copy_ users from one group into another, but does not give us an easy way to create rules that _automatically remove_ users from Okta groups. So the plan is indeed to remove all ENTRY_ONLY users from their corresponding USER groups, following this rollout. I've spoken to Alicia about this already to give her a heads up -- fortunately, this is a very small and manageable number of users to manually delete. After this, we ideally shouldn't have the problem again of needing to manually delete users from groups post-migration.

And to reiterate from a comment above, I also will be manually deleting all newly created group rules post-rollout. This is the only way to remove the rules without undoing the migration those rules performed upon creation.

File: backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java (hunk 8 above)
Comment (benwarfield-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java:387]: That's amazing.

File: backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java [hunk 12]
```diff
@@ -294,21 +296,114 @@ public void setUserIsActive(String username, Boolean active) {
...
-  public void createOrganization(Organization org) {
+  /**
+   * Iterates over all OrganizationRole's, creating new corresponding Okta groups for this
+   * organization where they do not already exist. For those OrganizationRole's that are in
+   * MIGRATION_DEST_ROLES and whose Okta groups are newly created, migrate all users from this org
+   * to those new Okta groups, where the migrated users are sourced from all pre-existing Okta
+   * groups for this organization. Separately, iterates over all facilities in this org, creating
+   * new corresponding Okta groups where they do not already exist. Does not perform any migration
+   * to these facility groups.
+   */
```
Comment (benwarfield-usds) [inline backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java:307]: 👍🏻

File: backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java [hunk 13]
```diff
@@ -294,21 +296,114 @@ public void setUserIsActive(String username, Boolean active) {
...
+          throw e;
+        }
+      }
+    }
+
+    if (migration && !migrationSourceGroupNames.isEmpty() && !migrationDestGroupIds.isEmpty()) {
+      String ruleName = generateMigrationGroupRuleName(externalId, MIGRATION_DEST_ROLES);
+      String ruleExpression = generateMigrationGroupRuleExpression(migrationSourceGroupNames);
+      try {
+        // https://developer.okta.com/docs/reference/api/groups/#group-rule-operations
```
Comment (benwarfield-usds) [inline backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java:374]: 🔗 🎉

File: backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java [hunk 14]
```diff
@@ -294,21 +296,114 @@ public void setUserIsActive(String username, Boolean active) {
...
+      }
     }
 
     _app.update();
   }
 
+  public void createOrganization(Organization org) {
+    createOrganization(org, Set.of(), false);
+  }
+
```
Comment (benwarfield-usds) [inline backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java:406]:
```suggestion
```
Interface takes care of this now, right?

**Review by benwarfield-usds — APPROVED** [review]
Tiny fit and finish suggestions, but LGTM overall—I am the opposite of envious of the process of figuring out all these API quirks, so thank you for all that slog!

File: backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java (hunk 14 above)
Comment (jeremyzitomer-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java:406]: well that would seem to make sense, yes 😆
